internal/metrics/metrics.go

```diff
@@ -0,0 +1,199 @@
+
+	opts.RetentionDuration = int64(DefaultRetentionDuration / time.Millisecond)
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	err := t.db.Close()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	err = os.RemoveAll(metricsDir)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (t *TSDB) GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error) {
+	log.Tracef("Getting metrics for %v - %v", tMin, tMax)
+	q, err := t.getDB().Querier(nil, toDbTime(tMin), toDbTime(tMax))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = q.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	matchers, err := getAllLabelsMatchers(q)
+	if err != nil {
+		return nil, err
+	}
+	seriesSet := q.Select(false, nil, matchers...)
+	if seriesSet.Err() != nil {
+		log.Error(seriesSet.Err())
+		return nil, seriesSet.Err()
+	}
+
+	return toSeries(seriesSet), nil
+}
+
+func toSeries(seriesSet storage.SeriesSet) []Series {
+	var allSeries []Series
+	for {
+		if !seriesSet.Next() {
+			break
+		}
+		s := seriesSet.At()
+		labelsMap := make(map[string]string)
+		for _, label := range s.Labels() {
+			labelsMap[label.Name] = label.Value
+		}
+		var dataPoints []DataPoint
+		for i := s.Iterator(); i.Next(); {
+			t, v := i.At()
+			dataPoints = append(dataPoints, DataPoint{Time: t, Value: v})
+		}
+		allSeries = append(allSeries, Series{
+			Labels:     labelsMap,
+			DataPoints: dataPoints,
+		})
+	}
+	return allSeries
+}
+
+func (t *TSDB) AddMetric(value float64, labelsMap map[string]string) error {
+	appender := t.getDB().Appender(context.Background())
+	_, err := appender.Append(0, labels.FromMap(labelsMap), toDbTime(time.Now()), value)
+	if err != nil {
+		return err
+	}
+
+	return appender.Commit()
+}
+
+func (t *TSDB) Update(config models.DeviceConfigurationMessage) error {
+	metricsConfiguration := config.Configuration.Metrics
+	maxBytes := DefaultMaxBytes
+	retentionDuration := DefaultRetentionDuration / time.Millisecond
+	if metricsConfiguration != nil && metricsConfiguration.Retention != nil {
+		maxBytes = int64(metricsConfiguration.Retention.MaxMib) * 1024 * 1024
+		retentionDuration = time.Duration(metricsConfiguration.Retention.MaxHours) * time.Hour / time.Millisecond
+	}
+	if int64(retentionDuration) == t.appliedOptions.RetentionDuration && maxBytes == t.appliedOptions.MaxBytes {
+		return nil
+	}
+	log.Infof("Metrics retention changed. MaxBytes [%v -> %v] RetentionDuration: [%d -> %d]", t.appliedOptions.MaxBytes, maxBytes, t.appliedOptions.RetentionDuration, retentionDuration)
+
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	dir := t.db.Dir()
+
+	err := t.db.Close()
```

Comment:
we are closing the DB while someone else might be writing to it. For example, AddMetric

Reply:
You're right; the Read lock is held not long enough and probably should span throughout the whole `AddMetric` and `GetMetricsForTimeRange` methods.

Reply:
Fixed.